monkey/infection_monkey/puppet/plugin_source_extractor.py

```diff
@@ -0,0 +1,113 @@
+        directory traversal and other relevant vulnerabilities, it is still advisable that you not
+        use this to extract archives/plugins from untrusted sources.
+
+        :param agent_plugin: An `AgentPlugin` to extract
+        :raises ValueError: If the agent's source code archive is potentially malicious or otherwise
+                            can not be extracted
+        """
+        destination = self._get_plugin_destination_directory(agent_plugin)
+        create_secure_directory(destination)
+
+        archive = TarFile(fileobj=io.BytesIO(agent_plugin.source_archive), mode="r")
+
+        # We check the entire archive to detect any malicious activity **before** we extract any
+        # files. This is a paranoid approach that prevents against partial extraction of malicious
+        # archives; either the whole archive is judged to be safe and extracted, or nothing is
+        # extracted.
+        detect_malicious_archive(destination, archive)
+        safe_extract(destination, archive)
+
+    def _get_plugin_destination_directory(self, agent_plugin: AgentPlugin) -> Path:
+        destination = (self.plugin_directory / agent_plugin.plugin_manifest.name).resolve()
+        self._detect_directory_traversal_in_plugin_name(destination)
+
+        return destination
+
+    def _detect_directory_traversal_in_plugin_name(self, destination: Path):
+        if self.plugin_directory.resolve() not in destination.resolve().parents:
+            raise ValueError(
+                f'Can not create directory "{destination}": directories must children of '
+                f'"{self.plugin_directory}"'
+            )
+
+
+UNSUPPORTED_MEMBER_TYPE_ERROR_MESSAGE = (
+    'The provided archive contains a file type other than "directory" or "regular"'
+)
+
+
+def safe_extract(destination_path: Path, archive: TarFile):
+    canonical_destination_path = destination_path.resolve()
+    for member in archive.getmembers():
+        # It's wise to perform this check for each member immediately before extraction, even if the
+        # archive has already been vetted by `detect_malicious_archive()`. This may help to mitigate
+        # TOCTOU attacks.
+        _detect_unsafe_extraction_conditions(canonical_destination_path, member)
+
+        if member.isdir():
+            create_secure_directory(canonical_destination_path / member.name)
+        elif member.isfile():
+            with open(canonical_destination_path / member.name, "wb") as f:
+                f.write(archive.extractfile(member).read())  # type: ignore [union-attr]
+        else:
+            # Note: This code should never run, since _detect_unsafe_extraction_conditions()
+            # should have already checked this. But this `else` clause is here for extra paranoia.
+            raise ValueError(UNSUPPORTED_MEMBER_TYPE_ERROR_MESSAGE)
+
+
+def detect_malicious_archive(destination_path: Path, archive: TarFile):
```

Comment:
Why is this not private?

Reply:
How about we move these to a separate `archive_utils.py` file?

Reply:
It's usually better to define things as close as possible to where they're used. I can make these private.